onyo/lib/onyo.py

```diff
@@ -1145,54 +1145,53 @@ def _read_asset(self, asset: Path) -> dict:
     def _set_sanitize(self, paths: Iterable[Union[Path, str]],
                       depth: Union[int, None]) -> list[Path]:
         """
-        Check and normalize a list of paths. Traverse directories and select all
-        assets from `paths` and return them as a list of absolute Paths.
+        Check and normalize a list of paths. Select all assets in the
+        repository that are relative to the given `paths` descending at most
+        `depth` directories. A `depth` of 0 descends without a limit.
         """
-        error_path_absent = []
-        error_path_protected = []
-        paths_to_set = []
-
         if depth and depth < 0:
             log.error(f"depth values must be positive, but is {depth}.")
             raise ValueError(f"depth values must be positive, but is {depth}.")
 
-        for p in paths:
-            full_path = Path(self.opdir, p).resolve()
-
-            # paths must exist
-            if not full_path.exists():
-                error_path_absent.append(p)
-                continue
-
-            # protected paths
-            if self._is_protected_path(full_path):
-                error_path_protected.append(p)
-                continue
-
-            if full_path.is_dir():
-                paths_to_set += self._select_assets_from_directory(full_path, depth)
-            else:
-                paths_to_set.append(full_path)
+        paths = {Path(p) for p in paths}
+
+        # TODO: This is superfluous due to using self.assets to find valid
+        #   set/unset targets -- it'll never hit non-existing or protected
+        #   paths
+        invalid_paths = {p for p in paths if not p.exists()}
+        if invalid_paths:
+            msg = (
```

Comment:
Because the next line uses the message in an error log, and the line after that uses it again to raise an exception. Your suggested approach means the message has to be written twice, risking potential redundancy when one message gets updated but the other is forgotten.

Reply:
Makes sense. Somehow I managed to miss that. ;-)